### PR TITLE
MODFIN-82 Create awaiting-payment API

### PR DESCRIPTION
**Pull request by grabsefx**
## Purpose
https://issues.folio.org/browse/MODFIN-82

## Approach
* API defined and implemented (RAML/ModuleDescriptor/code)
* Unit tests are updated

#### TODOS and Open Questions

* POST method implemented with response code "204 - No Content" as for no entities were created as result. Code 201 described in [wiki](https://wiki.folio.org/display/FOLIJET/Finance+Data+Model#FinanceDataModel-APIs)
* added  permission "finance.order-transaction-summaries.item.post" to "finance.all" list. It seems like it was missed in the one of previous PRs

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/rest/helper/AwaitingPaymentHelper.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  private Transaction modifyTransaction(Transaction transaction, Double amountAwaitingPayment) {
+    Double currentAwaitingPaymentAmount = transaction.getEncumbrance().getAmountAwaitingPayment();
+    Double currentAmountExpended = transaction.getEncumbrance().getAmountExpended();
+    String currency = transaction.getCurrency();
+
+    transaction.getEncumbrance()
+      .setAmountAwaitingPayment(MoneyUtils.sumDoubleValues(currentAwaitingPaymentAmount, amountAwaitingPayment, currency));
+    transaction.getEncumbrance()
+      .setAmountExpended(MoneyUtils.subtractDoubleValues(currentAmountExpended, amountAwaitingPayment, currency));
+    return transaction;
```
Comment (KVupp) [inline src/main/java/org/folio/rest/helper/AwaitingPaymentHelper.java]: The "releaseEncumbrance" field from the request, also has to be assigned to the encumbrance

File: src/main/java/org/folio/rest/helper/AwaitingPaymentHelper.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+  private Transaction modifyTransaction(Transaction transaction, Double amountAwaitingPayment) {
+    Double currentAwaitingPaymentAmount = transaction.getEncumbrance().getAmountAwaitingPayment();
+    Double currentAmountExpended = transaction.getEncumbrance().getAmountExpended();
+    String currency = transaction.getCurrency();
+
+    transaction.getEncumbrance()
+      .setAmountAwaitingPayment(MoneyUtils.sumDoubleValues(currentAwaitingPaymentAmount, amountAwaitingPayment, currency));
+    transaction.getEncumbrance()
+      .setAmountExpended(MoneyUtils.subtractDoubleValues(currentAmountExpended, amountAwaitingPayment, currency));
```
Comment (KVupp) [inline src/main/java/org/folio/rest/helper/AwaitingPaymentHelper.java]: the expended amount changes only when making a payment. This API just changes the awaiting payment amount, expended can be removed here

File: src/main/java/org/folio/rest/impl/AwaitingPaymentApi.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+  @Override
+  @Validate
+  public void postFinanceAwaitingPayment(String lang, AwaitingPayment awaitingPayment, Map<String, String> okapiHeaders,
+      Handler<AsyncResult<Response>> handler, Context vertxContext) {
+    AwaitingPaymentHelper helper = new AwaitingPaymentHelper(okapiHeaders, vertxContext, lang);
+    helper.moveToAwaitingPayment(awaitingPayment)
+      .thenAccept(v -> handler.handle(succeededFuture(helper.buildNoContentResponse())))
+      .exceptionally(t -> {
+        handler.handle(succeededFuture(helper.buildErrorResponse(t)));
```
Comment (KVupp) [inline src/main/java/org/folio/rest/impl/AwaitingPaymentApi.java]: there is a helper utils method for this

File: ramls/awaiting-payment.raml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+traits:
+  validate: !include raml-util/traits/validation.raml
+  language: !include raml-util/traits/language.raml
+
+/finance/awaiting-payment:
+  displayName: Finance awaiting payment
+  description: Finance awaiting payment APIs
+  is: [language]
+  post:
```
Comment (javalkar-varun) [inline ramls/awaiting-payment.raml]: If this API only has POST, please reuse `rtypes/post-json-201.raml`

Comment (javalkar-varun) [reply ramls/awaiting-payment.raml]: Never mind, API returns 204

File: src/main/java/org/folio/rest/helper/AwaitingPaymentHelper.java (hunk 1 above)
Comment (grabsefx) [reply src/main/java/org/folio/rest/helper/AwaitingPaymentHelper.java]: FIxed. Encumbrance status updates depending on "releaseEncumbrance" property

File: src/main/java/org/folio/rest/helper/AwaitingPaymentHelper.java (hunk 2 above)
Comment (grabsefx) [reply src/main/java/org/folio/rest/helper/AwaitingPaymentHelper.java]: removed

File: src/main/java/org/folio/rest/helper/AwaitingPaymentHelper.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import org.folio.rest.jaxrs.model.AwaitingPayment;
+import org.folio.rest.jaxrs.model.Encumbrance;
+import org.folio.rest.jaxrs.model.Transaction;
+import org.folio.rest.util.MoneyUtils;
+
+import io.vertx.core.Context;
+
+public class AwaitingPaymentHelper extends AbstractHelper {
+
+  public AwaitingPaymentHelper(Map<String, String> okapiHeaders, Context ctx, String lang) {
```
Comment (pautau-s) [inline src/main/java/org/folio/rest/helper/AwaitingPaymentHelper.java]: Awaiting-payment is part of Transactions API and it just a replacement for the PUT /finance/encumbrance/<id>, so I think all this logic can be in Transaction helper.

Comment (grabsefx) [reply src/main/java/org/folio/rest/helper/AwaitingPaymentHelper.java]: agree

**Pull request merged**